lib/pos_staking_rewards.go

```diff
@@ -177,39 +182,51 @@ func (bav *UtxoView) computeStakerRewardAndValidatorCommission(
 	return stakerRewardNanos.Uint64(), validatorCommissionNanos.Uint64(), nil
 }
 
-func (bav *UtxoView) distributeStakingReward(validatorPKID *PKID, stakerPKID *PKID, rewardNanos uint64) error {
+func (bav *UtxoView) distributeStakingReward(validatorPKID *PKID, stakerPKID *PKID, rewardNanos uint64) (*UtxoOperation, error) {
 	// Fetch the staker's latest StakeEntry.
 	stakeEntry, err := bav.GetStakeEntry(validatorPKID, stakerPKID)
 	if err != nil {
-		return errors.Wrapf(err, "distributeStakingReward: problem fetching staker's StakeEntry: ")
+		return nil, errors.Wrapf(err, "distributeStakingReward: problem fetching staker's StakeEntry: ")
 	}
 
 	// At this point, there are three possible cases:
 	// 1. The stake entry still exists and wants to restake their rewards.
 	// 2. The stake entry still exists and does not want to restake their rewards.
 	// 3. The stake entry has unstaked since the snapshot was taken.
 
+	var utxoOperation *UtxoOperation
 	// For case 1, we distribute the rewards by adding them to the staker's staked amount.
 	if stakeEntry != nil && stakeEntry.RewardMethod == StakingRewardMethodRestake {
+		validatorEntry, err := bav.GetValidatorByPKID(stakeEntry.ValidatorPKID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "distributeStakingReward: problem fetching validator entry: ")
+		}
+		utxoOperation = &UtxoOperation{
+			Type:                 OperationTypeStakeDistribution,
+			PrevStakeEntries:     []*StakeEntry{stakeEntry.Copy()},
+			PrevValidatorEntry:   validatorEntry.Copy(),
+			StakeAmountNanosDiff: rewardNanos,
+		}
 		stakeEntry.StakeAmountNanos = uint256.NewInt().Add(stakeEntry.StakeAmountNanos, uint256.NewInt().SetUint64(rewardNanos))
 		bav._setStakeEntryMappings(stakeEntry)
-
-		return nil
+		validatorEntry.TotalStakeAmountNanos = uint256.NewInt().Add(validatorEntry.TotalStakeAmountNanos, uint256.NewInt().SetUint64(rewardNanos))
+		bav._setValidatorEntryMappings(validatorEntry)
```

Comment:
Oh interesting. So it looks like we had a bug before whereby we weren't updating the TotalStakeAmountNanos on a validator when restaking right?

Reply:
That's correct! Found it while I was in here.

Reply:
And this reminds me that we also need to update the global total stake amount nanos value

Reply:
Can we add a sanity-check that runs at the end of every Nth RunEpochCompleteHook and either logs an error or panics if these values differ? Ideally the N value could be adjusted by ParamUpdater so that we can start it =1 then make it bigger to increase efficiency.

Reply:
Honestly even if it just ran every time to start, that would probably help us catch bugs, and we can remove it manually if it ends up bottlenecking us.

Reply:
@diamondhands0 @lazynina we no longer have a global state amount nanos Badger index. So, we don't need to and wouldn't be able to run the cross-validation suggested here.